### ISSUE TITLE
Waiting on CPU resources while only one user in system

### ISSUE THREAD

**Issue by e-carlin** [radiasoft/sirepo#4660]
I've been working on https://github.com/radiasoft/sirepo/pull/4651#issuecomment-1198559148 trying to trace down why there is contention between querying for scans (`Op(kind=sequential, name=analysis)`) and running analysis (`Op(kind=sequential, name=run)`). In the logs I'm seeing messages like:
```
./sirepo/job_supervisor.py:121:alloc op=_Op(analysis, QwDh)  situation=Waiting for CPU resources
```

I don't think there is ever a scenario where there is one user in the system, one configured cpu_slot (of each kind), and there will be contention causing `Waiting for CPU resources`. The single user should always have the cpu slot so should always be able to use it. 

I think I've narrowed the issue down to how we use `sirepo.job_supervisor.SlotProxy`. Each time an op is created we call [sr_slot_proxy](https://github.com/radiasoft/sirepo/blob/6c9c91a2f77e3a6e6e087e27dbb4a29b431cbf21/sirepo/job_supervisor.py#L1064_ and a new SRSlotProxy object is instantiated. I think the problem is that a new SRSlotProxy object should not be created. They contain state (`self._value`) and that state matters between ops. In the case of cpu_slot if we already have a slot, then `self._value` should be `not None` and we won't get into this contention situation. But, as it is currently written each op gets a different instance of SRSlotProxy so no state is shared so each op is waiting on its own cpu slot. 

I still need to confirm whether this logic applies to op_slot's and run_dir_slot's. As well as whether this is impacting just LocalDrivers or all drivers.

@robnagler I would appreciate if you put a little thinking into this and confirmed that my logic makes sense. I'm concerned I'm overlooking some detail and the system is actually operating correctly.

**Comment by e-carlin** [comment]
If it helps, I've made a branch `slot-proxy` that has additional logging that helped me trace things down.

**Comment by robnagler** [comment]
> querying for scans (Op(kind=sequential, name=analysis)) and running analysis (Op(kind=sequential, name=run))
> [snip]
> I don't think there is ever a scenario where there is one user in the system, one configured cpu_slot (of each kind)

The two ops are both `sequential` (same kind) so they will contend on `cpu_slot_q` for `docker` and `local`, but not `sbatch`, which is not restricted by the number of cpus. There is only one core in a docker container, which is a real constraint. An `analysis op` does take time.

I totally forgot how this worked so I spent time documenting in 08e8ce05. More could be done.

**Comment by e-carlin** [comment]
Thanks for looking into it! I appreciate the added documentation. 

Unfortunately there is still something I'm not seeing. I'm sorry to have to ask but can you help shed some light?

A couple constraints to help shape the conversation:
- 1 user in the system
- queue size of 1 for all slots (op_slots, cpu_slots, run_dir_slots)
- local driver only

> The two ops are both sequential (same kind) so they will contend on cpu_slot_q

My understanding is that cpu_slot_q is for contention **between users**. That is, if there is only one user in the system, one sequential cpu_slot, and one parallel cpu_slot (as there was above) then we should never contend for the cpu_slots. The single user in the system should always have the cpu_slots. But, where there should be contention is on op_slot. Those jobs above are of different "names". One is analysis and the other is run. That means they will never contend for the same op_slot and they should be allowed to run at the same time using the same cpu_slot. But, jobs of the same name (ex another analysis job) will contend for the op_slot.

In my example there was only one user in the system so I believe we should never see a message like 'Waiting for CPU resources' (which I saw).

Phrased another way, do we allow an analysis (always sequential) and a sequential run to run at the same time on the same agent? My understanding of the system was yes we do allow it. But, I very well could be wrong.

If we don't allow it then I'm struggling to see why we have an op_slot_q. I believe everything could just wait on cpu_slot_q. 

Thanks for your help.

**Comment by robnagler** [comment]
> My understanding is that cpu_slot_q is for contention between users.

No. a cpu slot is a contention between ops (processes) that consume cpu. Just like a run dir slot is contention between ops that might read/write the run_dir.

System sizing is extremely important. It was something that was drilled into me at Tandem, but I had some experience with it before. A single core can only execute one process, that's it. Most of what we do is compute intensive, even analysis. 

We have talked before about more kinds of agents, e.g. to run sequential long running jobs. We could add that type of agent, or we could have one agent with two cores: one for "long" sequential runs and one for other ops (analysis, io, etc.). This would require more sophisticated queueing either way.

**Comment by robnagler** [comment]
> If we don't allow it then I'm struggling to see why we have an op_slot_q. I believe everything could just wait on cpu_slot_q.

op_slot_q is about types of ops: analysis, io, run, etc. Even if we had two cores (two cpu slots), we might only want one op type to run at a time.

**Comment by e-carlin** [comment]
> No. a cpu slot is a contention between ops (processes) that consume cpu

Ah, thanks!

> op_slot_q is about types of ops: analysis, io, run, etc. Even if we had two cores (two cpu slots), we might only want one op type to run at a time.

Ok, that's making more sense. When you say "one op type to run at a time" do you mean one op of *each* type or one op of *all* types?

If the answer is one op of *each* type: why do we want to restrict the system in that way?

If the answer is one op of *all* types: why not just let the cpu_slot_q control those resources?

Another question that comes up relates to docker. We constrain cpu resources for our containers. Is it a problem if a sequential container is running, say, an analysis and a run at the same time? Could that be a problem with the "system sizing" you mentioned above?

Thanks again for your patience in helping me understand.

**Comment by e-carlin** [comment]
Taking a break let me think about this more clearly. It is for each op. We want to constrain the system in that way to prevent users from gobbling up all available cpu resources for their jobs. 

Thanks again for walking through it.

**Comment by robnagler** [comment]
> If the answer is one op of *all* types: 

It is "all". We would want to reserve one core for long running ops (currently only "run"), and the other core for "short" ops e.g. analysis or io.

> why not just let the cpu_slot_q control those resources?

To avoid the problem you are seeing. Consider the problem of all reports starting up at the same time or two browser windows. We want to give priority to analysis and io jobs without having to kill run jobs. We don't want the gui to have to manage the queue like it does now, because it doesn't solve the problem of two browser windows. The server has the resources it is controlling. The browser should simply initiate all requests and the server should queue them.

The fundamental issue we have is that we are running arbitrary end-user code. This is not a problem that most systems have. This means normal global queuing algorithms don't work. 

That being said, we might be able to get away with running analysis and io jobs in another type of docker container. However, we would have to make sure we maintained control of the code that is executing in those containers. This is a hard problem given that we are running arbitrary "codes" sometimes to do analysis. This is like running end-user code in that we have no control over the security. 

One solution to this particular problem would be to dynamically, logically "mount" user volumes inside a container. If the name of the directory kept changing and we didn't let the op run too long, it could be possible to "ensure" that nothing bad happens during an op to other user data. 

Another solution would be to approve particular op-parameter combinations. For example, if we knew we were only using matplotlib, hdf5, etc. for a particular analysis or io or even stateless/ful compute, we could send those operations to a "safe op" queue, which would allow much more parallelism.

With current system loads and the cost of our hardware, it is a bit hard to justify all this extra, non-scalable work. Rather, we can just oversize the containers to get more responsiveness on io and analysis ops.

**Comment by e-carlin** [comment]
@robnagler and I had a productive conversation about this topic this morning. Below are the takeaways:

Ultimately, we agreed that there was a bug (at least one) in the way we are allocating resources. We can get into a scenario where all of the cpu_slots are allocated but there are fewer docker containers (constrained on the number of cpu's they can use) than the number of cores on the system. With our current queues we don't have enough of a handle on what agents are using what resources and what ops those agents are running that require resources.

We couldn't exactly agree on what cpu_slots are currently for and how they should be allocated so we decided to get rid of name/concept of cpu_slots and instead add two new slot types to the system:

1. active_agent_slots: These control which agents can be actively running op(s) that are in _CPU_SLOT_OPS. They have some relationship to the number of cores on a system. For example, if there are 4 cores on a machine and we have 4 agents that are constrained to use one core then we may have 4 active_agent_slots and assign one to each of the agents. Or we may constrain each agent to two cores, have 2 active_agent_slots in the system, and assign one to 2 of the agents while the other two wait. Or we may have some other scenario where the number of active_agent_slots * number of cores that each container is constrained to is != number of cores on the host but instead has some other relationship to it. In any case for an agent to be running an op in _CPU_SLOT_OPS they must be assigned one (and only one) active_agent_slot.
2. compute_bound_slots: These control how many ops in _CPU_SLOT_OPS an individual agent can run at the same time. They are local to an agent. So, if compute_bound_slots is set to 2 and an agent is running a run op and an analysis op then if another run or analysis op comes in it will have to wait until one of the two ops in front of it completes (assuming the op has already gained any earlier resources like op_slot_q). 

We also discussed the need for another queue to handle idle agents. That is, if a host was fully subscribed then an agent that wanted to start would have to wait in this queue until an existing agent became idle and could be garbage collected.

Finally, we settled on this not being an immediate priority to fix. The system is working as is so is ok for now. In the future when fixing these problems we may re-write some of the logic to fix this and add the concept of things like "long running sequential agent" vs "short running sequential agent" and having multiple of the same type of agents per user. 

@robnagler does that all sound right?

**Comment by robnagler** [comment]
> active_agent_slots: These control which agents can be actively running op(s) that are in _CPU_SLOT_OPS

I think this is any op, not just _CPU_SLOT_OPS (which should be renamed _COMPUTE_BOUND_OPS).

An agent is active, if it is running anything. An agent is idle, if it is not running anything.

**Comment by e-carlin** [comment]
> This assumes that the number of op slots available to analysis jobs > 1. That is currently not the case, but we need to make it more flexible for raydata.

Agreed. I tried to implying by the parenthetical "(assuming the op has already gained any earlier resources like op_slot_q)"

> The "active_agent_slots" is what controls global resources and "op_slots" controls local resources.

That makes sense to me. I don't have intuition on whether it will be a problem to have at least one of each op type able to run concurrently. But, assuming we size the container resources constraint properly it shouldn't be a problem.

> When we size the system, we ensure the length of these queues will not oversubscribe the agent.

Sounds good.

> It is important to note that active_agent_slots needs to be constrained in all cases, even sbatch

Make sense.
